esx_service/vmdk_ops.py

```diff
@@ -691,6 +709,38 @@ def getStatusAttached(vmdk_path):
         uuid = None
     return attached, uuid, attach_as
 
+def handle_stale_attach(vmdk_path, kv_uuid):
+       '''
+       Clear volume state for cases where the VM that attached the disk
+       earlier is powered off or removed, detach the disk from the VM
+       if its powered off.
+       '''
+       cur_vm = findVmByUuid(kv_uuid)
+
+       if cur_vm:
+          # Detach the disk only if VM is powered off
+          if cur_vm.runtime.powerState == VM_POWERED_OFF:
+             logging.warning("Detaching disk %s from VM(powered off) - %s\n",
+                             vmdk_path, cur_vm.config.name)
+             device = findDeviceByPath(vmdk_path, cur_vm)
+             if device:
+                msg = disk_detach(vmdk_path, cur_vm, device)
+                if msg:
+                   msg += " failed to detach disk {0} from VM={1}.".format(vmdk_path,
+                                                                           cur_vm.config.name)
+                   return err(msg)
+             else:
+                logging.warning("Failed to find disk %s in powered off VM - %s\n",
```

Comment:
Nit: this probably means corrupted metadata, so I think message should be a bit more descriptive. Maybe add "Resetting volume metadata" ?